### PR TITLE
DAG: Implement promotion for strict_fp_round

### PR DESCRIPTION
**Pull request by arsenm**
Needs an AMDGPU hack to get the selection to work. The ordinary
variant is custom lowered through an almost equivalent target node
that would need a strict variant for additional known bits optimizations.

### CODE REVIEW

**Comment by arsenm** [comment]
ping

File: llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp [hunk 1]
```diff
@@ -2621,6 +2642,29 @@ SDValue DAGTypeLegalizer::PromoteFloatRes_FP_ROUND(SDNode *N) {
...
+  EVT OpVT = Op->getValueType(0);
+  EVT NVT = TLI.getTypeToTransformTo(*DAG.getContext(), N->getValueType(0));
+  EVT IVT = EVT::getIntegerVT(*DAG.getContext(), VT.getSizeInBits());
+
+  // Round promoted float to desired precision
+  SDValue Round = DAG.getNode(GetPromotionOpcodeStrict(OpVT, VT), DL,
+                              DAG.getVTList(IVT, MVT::Other), Chain, Op);
+  // Promote it back to the legal output type
+  SDValue Res =
+      DAG.getNode(GetPromotionOpcodeStrict(VT, NVT), DL,
```
Comment (andykaylor) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:2662]: GetPromotionOpcode() is used in a lot more places. Are those not needed for strict or are they coming in later patches?

**Review by andykaylor — APPROVED** [review]
This looks good to me. I've added Phoebe Wang as a reviewer because she knows the SelectionDAG code better than I do and might want to take a look. I'm assuming you have reviewed the AMDGPU-specific parts of this with someone who knows that architecture.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:2662]: They all need updating at some point

File: llvm/lib/Target/AMDGPU/SIInstructions.td [hunk 2]
```diff
@@ -1097,7 +1097,7 @@ def : Pat <
 multiclass f16_fp_Pats<Instruction cvt_f16_f32_inst_e64, Instruction cvt_f32_f16_inst_e64> {
   // f16_to_fp patterns
   def : GCNPat <
-    (f32 (f16_to_fp i32:$src0)),
+    (f32 (any_f16_to_fp i32:$src0)),
```
Comment (phoebewang) [inline llvm/lib/Target/AMDGPU/SIInstructions.td:1100]: Should we replace more `f16_to_fp` to `any_f16_to_fp` in this file and other target files?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstructions.td:1100]: Yes, with appropriate tests added alongside it

File: llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp [hunk 3]
```diff
@@ -2621,6 +2642,29 @@ SDValue DAGTypeLegalizer::PromoteFloatRes_FP_ROUND(SDNode *N) {
   return DAG.getNode(GetPromotionOpcode(VT, NVT), DL, NVT, Round);
 }
 
+// Explicit operation to reduce precision.  Reduce the value to half precision
+// and promote it back to the legal type.
+SDValue DAGTypeLegalizer::PromoteFloatRes_STRICT_FP_ROUND(SDNode *N) {
```
Comment (phoebewang) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:2647]: Why we need this given we have a handler here https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp#L2996-L3003

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:2647]: There are two promotion legalizers, Promote and SoftPromoteHalf. X86 and RISC-V use SoftPromoteHalf. Other targets use Promote. Promote doesn't convert back to fp16 in the right places in my opinion. If I recall correctly Promote doesn't convert to fp16 between two promoted fadds for example so it keeps extra precision.

**Pull request merged**